include/ginkgo/extensions/kokkos/types.hpp

```diff
@@ -0,0 +1,285 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GINKGO_EXTENSIONS_KOKKOS_TYPES_HPP
+#define GINKGO_EXTENSIONS_KOKKOS_TYPES_HPP
+
+#include <Kokkos_Complex.hpp>
+
+
+#include <Kokkos_Core.hpp>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+namespace ext {
+namespace kokkos {
+namespace detail {
+
+
+/**
+ * Maps arithmetic types to corresponding Kokkos types.
+ *
+ * @tparam T  An arithmetic type.
+ */
+template <typename T>
+struct value_type_impl {
+    using type = T;
+};
+
+template <typename T>
+struct value_type_impl<const T> {
+    using type = const typename value_type_impl<T>::type;
+};
+
+template <typename T>
+struct value_type_impl<std::complex<T>> {
+    using type = Kokkos::complex<T>;
+};
+
+
+template <typename T>
+struct value_type {
+    using type = typename value_type_impl<T>::type;
+
+    static_assert(sizeof(std::decay_t<T>) == sizeof(std::decay_t<type>),
+                  "Can't handle C++ data type and corresponding Kokkos "
+                  "type with mismatching type sizes.");
+#if GINKGO_EXTENSION_KOKKOS_CHECK_TYPE_ALIGNMENT
```

Comment:
is this optional check because user might only use real type?

Reply:
That is a bit more about lazyness. By default kokkos doesn't use the same alignment as `std::complex`, but you can't change that when installing through spack. Since this doesn't seem to lead to issues usually, I've added this option.